File: Assets/Scripts/DouyinSequenceRegistry.lua
---@var PendulumGroup          :DouyinScript
---@var ExtendGroup            :DouyinScript
---@var BallGroup              :DouyinScript
---@var OrbitGroup             :DouyinScript
---@var GearGroup              :DouyinScript
---@var FxBridge               :DouyinScript
---@var VertexLightManager     :DouyinScript
---@var EnableValidationLog    :bool = true
---@end

local function Log(message)
    if EnableValidationLog then
        print("[DouyinSequenceRegistry] " .. tostring(message))
    end
end

local function CallScript(target, funcName, ...)
    if target == nil or target.script == nil then
        return nil
    end

    local fn = target.script[funcName]
    if fn == nil then
        return nil
    end

    return fn(...)
end

local function ValidateScript(target, label)
    if target == nil then
        Log(label .. " is nil")
        return false
    end
    if target.script == nil then
        Log(label .. ".script is nil")
        return false
    end
    return true
end

function Validate()
    local ok = true
    ok = ValidateScript(PendulumGroup, "PendulumGroup") and ok
    ok = ValidateScript(ExtendGroup, "ExtendGroup") and ok
    ok = ValidateScript(BallGroup, "BallGroup") and ok
    ok = ValidateScript(OrbitGroup, "OrbitGroup") and ok
    ok = ValidateScript(GearGroup, "GearGroup") and ok
    if FxBridge ~= nil then
        ok = ValidateScript(FxBridge, "FxBridge") and ok
    end
    if VertexLightManager ~= nil then
        ok = ValidateScript(VertexLightManager, "VertexLightManager") and ok
    end
    return ok
end

function CollectAll()
    CallScript(PendulumGroup, "CollectUnits")
    CallScript(ExtendGroup, "CollectUnits")
    CallScript(BallGroup, "CollectUnits")
    CallScript(OrbitGroup, "CollectUnits")
    CallScript(GearGroup, "CollectUnits")
    CallScript(VertexLightManager, "CollectLights")
end

function ResetAllToBase()
    CallScript(PendulumGroup, "ResetToBase")
    CallScript(ExtendGroup, "ResetToBase")
    CallScript(BallGroup, "ResetToBase")
    CallScript(OrbitGroup, "ResetToBase")
    CallScript(GearGroup, "ResetToBase")
    CallScript(FxBridge, "ResetToBase")
    CallScript(VertexLightManager, "ResetToBase")
end

function BeginPendulums()
    CallScript(PendulumGroup, "BeginForward")
end

function TickPendulums(deltaTime)
    CallScript(PendulumGroup, "Tick", deltaTime)
end

function ArePendulumsComplete()
    return CallScript(PendulumGroup, "IsComplete") == true
end

function BeginExtendForward()
    CallScript(ExtendGroup, "BeginForward")
end

function BeginExtendBackward()
    CallScript(ExtendGroup, "BeginBackward")
end

function TickExtenders(deltaTime)
    CallScript(ExtendGroup, "Tick", deltaTime)
end

function AreExtendersComplete()
    return CallScript(ExtendGroup, "IsComplete") == true
end

function BeginBallForward()
    CallScript(BallGroup, "BeginForward")
end

function BeginBallBackward()
    CallScript(BallGroup, "BeginBackward")
end

function TickBalls(deltaTime)
    CallScript(BallGroup, "Tick", deltaTime)
end

function AreBallsComplete()
    return CallScript(BallGroup, "IsComplete") == true
end

function BeginOrbit()
    CallScript(OrbitGroup, "BeginForward")
end

function TickOrbit(deltaTime)
    CallScript(OrbitGroup, "Tick", deltaTime)
end

function IsOrbitComplete()
    return CallScript(OrbitGroup, "IsComplete") == true
end

function GetOrbitSpeedFactor()
    local speedFactor = CallScript(OrbitGroup, "GetSpeedFactor")
    if speedFactor == nil then
        return 0
    end
    return speedFactor
end

function GetOrbitProgress01()
    local progress = CallScript(OrbitGroup, "GetProgress01")
    if progress == nil then
        return 0
    end
    return progress
end

function GetOrbitDuration()
    local duration = CallScript(OrbitGroup, "GetTotalDuration")
    if duration == nil then
        return 0
    end
    return duration
end

function TickGears(deltaTime, driveFactor)
    CallScript(GearGroup, "Tick", deltaTime, driveFactor)
end

function BeginOrbitFx(totalDuration)
    CallScript(FxBridge, "BeginOrbitFx", totalDuration)
end

function TickOrbitFx(deltaTime, progress01)
    CallScript(FxBridge, "TickOrbitFx", deltaTime, progress01)
end

function CompleteOrbitFx()
    CallScript(FxBridge, "CompleteOrbitFx")
end

function GetOrbitNightBlend01(progress01)
    local blend = CallScript(FxBridge, "GetOrbitNightBlend01", progress01)
    if blend == nil then
        return 0
    end
    return blend
end

function GetCurrentNightBlend01()
    local blend = CallScript(FxBridge, "GetCurrentNightBlend01")
    if blend == nil then
        return 0
    end
    return blend
end

function TickVertexLights(deltaTime)
    CallScript(VertexLightManager, "Tick", deltaTime)
end

function ApplyVertexLights()
    CallScript(VertexLightManager, "ApplyGlobals")
end

function SetVertexLightIntensity(lightIndex, intensity)
    CallScript(VertexLightManager, "SetLightIntensity", lightIndex, intensity)
end

function SetVertexLightIntensityScale(lightIndex, scale)
    CallScript(VertexLightManager, "SetLightIntensityScale", lightIndex, scale)
end

function FadeVertexLightIntensity(lightIndex, targetIntensity, duration)
    CallScript(VertexLightManager, "FadeLightIntensity", lightIndex, targetIntensity, duration)
end

function FadeVertexLightIntensityScale(lightIndex, targetScale, duration)
    CallScript(VertexLightManager, "FadeLightIntensityScale", lightIndex, targetScale, duration)
end

function SetAllVertexLightIntensityScale(scale)
    CallScript(VertexLightManager, "SetAllIntensityScale", scale)
end

function FadeAllVertexLightIntensityScale(targetScale, duration)
    CallScript(VertexLightManager, "FadeAllIntensityScale", targetScale, duration)
end

File: Assets/Scripts/DouyinSequenceRegistry.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using UnityEngine;

public sealed class DouyinSequenceRegistry : MonoBehaviour
{
    public DouyinSequencePendulumGroup PendulumGroup;
    public DouyinSequenceMoveGroup ExtendGroup;
    public DouyinSequenceRotateGroup BallGroup;
    public DouyinSequenceOrbitGroup OrbitGroup;
    public DouyinSequenceGearGroup GearGroup;
    public DouyinSequenceFxBridge FxBridge;
    public DouyinVertexLightManager VertexLightManager;
    public DouyinSequenceAudioBridge AudioBridge;
    public bool EnableValidationLog = true;

    public bool Validate()
    {
        bool ok = true;
        ok = ValidateReference(PendulumGroup, "PendulumGroup") && ok;
        ok = ValidateReference(ExtendGroup, "ExtendGroup") && ok;
        ok = ValidateReference(BallGroup, "BallGroup") && ok;
        ok = ValidateReference(OrbitGroup, "OrbitGroup") && ok;
        ok = ValidateReference(GearGroup, "GearGroup") && ok;
        if (FxBridge != null) ok = ValidateReference(FxBridge, "FxBridge") && ok;
        if (VertexLightManager != null) ok = ValidateReference(VertexLightManager, "VertexLightManager") && ok;
        if (AudioBridge != null) ok = ValidateReference(AudioBridge, "AudioBridge") && ok;
        return ok;
    }

    public void CollectAll()
    {
        PendulumGroup?.CollectUnits();
        ExtendGroup?.CollectUnits();
        BallGroup?.CollectUnits();
        OrbitGroup?.CollectUnits();
        GearGroup?.CollectUnits();
        VertexLightManager?.CollectLights();
    }

    public void ResetAllToBase()
    {
        PendulumGroup?.ResetToBase();
        ExtendGroup?.ResetToBase();
        BallGroup?.ResetToBase();
        OrbitGroup?.ResetToBase();
        GearGroup?.ResetToBase();
        FxBridge?.ResetToBase();
        VertexLightManager?.ResetToBase();
        AudioBridge?.ResetToBase();
    }

    public void BeginPendulums() => PendulumGroup?.BeginForward();
    public void TickPendulums(float deltaTime) => PendulumGroup?.Tick(deltaTime);
    public bool ArePendulumsComplete() => PendulumGroup == null || PendulumGroup.IsComplete();
    public void BeginExtendForward() => ExtendGroup?.BeginForward();
    public void BeginExtendBackward() => ExtendGroup?.BeginBackward();
    public void TickExtenders(float deltaTime) => ExtendGroup?.Tick(deltaTime);
    public bool AreExtendersComplete() => ExtendGroup == null || ExtendGroup.IsComplete();
    public void BeginBallForward() => BallGroup?.BeginForward();
    public void BeginBallBackward() => BallGroup?.BeginBackward();
    public void TickBalls(float deltaTime) => BallGroup?.Tick(deltaTime);
    public bool AreBallsComplete() => BallGroup == null || BallGroup.IsComplete();
    public void BeginOrbit() => OrbitGroup?.BeginForward();
    public void TickOrbit(float deltaTime) => OrbitGroup?.Tick(deltaTime);
    public bool IsOrbitComplete() => OrbitGroup == null || OrbitGroup.IsComplete();
    public float GetOrbitSpeedFactor() => OrbitGroup != null ? OrbitGroup.GetSpeedFactor() : 0f;
    public float GetOrbitProgress01() => OrbitGroup != null ? OrbitGroup.GetProgress01() : 0f;
    public float GetOrbitDuration() => OrbitGroup != null ? OrbitGroup.GetTotalDuration() : 0f;
    public void TickGears(float deltaTime, float? driveFactor) => GearGroup?.Tick(deltaTime, driveFactor);
    public void BeginOrbitFx(float totalDuration) => FxBridge?.BeginOrbitFx(totalDuration);
    public void TickOrbitFx(float deltaTime, float progress01) => FxBridge?.TickOrbitFx(deltaTime, progress01);
    public void CompleteOrbitFx() => FxBridge?.CompleteOrbitFx();
    public float GetOrbitNightBlend01(float progress01) => FxBridge != null ? FxBridge.GetOrbitNightBlend01(progress01) : 0f;
    public float GetCurrentNightBlend01() => FxBridge != null ? FxBridge.GetCurrentNightBlend01() : 0f;
    public void TickVertexLights(float deltaTime) => VertexLightManager?.Tick(deltaTime);
    public void ApplyVertexLights() => VertexLightManager?.ApplyGlobals();
    public void SetVertexLightIntensity(int lightIndex, float intensity) => VertexLightManager?.SetLightIntensity(lightIndex, intensity);
    public void SetVertexLightIntensityScale(int lightIndex, float scale) => VertexLightManager?.SetLightIntensityScale(lightIndex, scale);
    public void FadeVertexLightIntensity(int lightIndex, float targetIntensity, float duration) => VertexLightManager?.FadeLightIntensity(lightIndex, targetIntensity, duration);
    public void FadeVertexLightIntensityScale(int lightIndex, float targetScale, float duration) => VertexLightManager?.FadeLightIntensityScale(lightIndex, targetScale, duration);
    public void SetAllVertexLightIntensityScale(float scale) => VertexLightManager?.SetAllIntensityScale(scale);
    public void FadeAllVertexLightIntensityScale(float targetScale, float duration) => VertexLightManager?.FadeAllIntensityScale(targetScale, duration);
    public void BeginSequenceAudio(string stateName) => AudioBridge?.BeginState(stateName);
    public void TickSequenceAudio(float deltaTime) => AudioBridge?.Tick(deltaTime);
    public void PauseSequenceAudio() => AudioBridge?.PauseAll();
    public void ResumeSequenceAudio() => AudioBridge?.ResumeActive();
    public void StopSequenceAudio() => AudioBridge?.StopAll();

    private bool ValidateReference(UnityEngine.Object target, string label)
    {
        if (target != null)
        {
            return true;
        }

        if (EnableValidationLog)
        {
            Debug.Log("[DouyinSequenceRegistry] " + label + " is nil", this);
        }
        return false;
    }
}

